server/src/main/java/org/opensearch/common/util/BytesRefHash.java

```diff
@@ -6,179 +6,291 @@
+     */
+    private LongArray table;
 
-    // Constructor with configurable capacity and default maximum load factor.
-    public BytesRefHash(long capacity, BigArrays bigArrays) {
-        this(capacity, DEFAULT_MAX_LOAD_FACTOR, bigArrays);
+    /**
+     * Underlying array to store the starting offsets of keys.
+     *
+     * <p>
+     * Terminology:
+     * <pre>
+     *   offsets[ordinal] = starting offset (inclusive)
+     *   offsets[ordinal + 1] = ending offset (exclusive)
+     * </pre>
+     */
+    private LongArray offsets;
+
+    /**
+     * Underlying byte array to store the keys.
+     *
+     * <p>
+     * Terminology: <code>keys[start...end] = key</code>
+     */
+    private ByteArray keys;
+
+    public BytesRefHash(final BigArrays bigArrays) {
+        this(DEFAULT_INITIAL_CAPACITY, DEFAULT_LOAD_FACTOR, DEFAULT_HASHER, bigArrays);
     }
 
-    // Constructor with configurable capacity and load factor.
-    public BytesRefHash(long capacity, float maxLoadFactor, BigArrays bigArrays) {
-        super(capacity, maxLoadFactor, bigArrays);
-        startOffsets = bigArrays.newLongArray(capacity + 1, false);
-        startOffsets.set(0, 0);
-        bytes = bigArrays.newByteArray(capacity * 3, false);
-        hashes = bigArrays.newIntArray(capacity, false);
-        spare = new BytesRef();
+    public BytesRefHash(final long initialCapacity, final BigArrays bigArrays) {
+        this(initialCapacity, DEFAULT_LOAD_FACTOR, DEFAULT_HASHER, bigArrays);
     }
 
-    // BytesRef has a weak hashCode function so we try to improve it by rehashing using Murmur3
-    // Feel free to remove rehashing if BytesRef gets a better hash function
-    private static int rehash(int hash) {
-        return BitMixer.mix32(hash);
+    public BytesRefHash(final long initialCapacity, final float loadFactor, final BigArrays bigArrays) {
+        this(initialCapacity, loadFactor, DEFAULT_HASHER, bigArrays);
     }
 
-    /**
-     * Return the key at <code>0 &lt;= index &lt;= capacity()</code>. The result is undefined if the slot is unused.
-     * <p>Beware that the content of the {@link BytesRef} may become invalid as soon as {@link #close()} is called</p>
-     */
-    public BytesRef get(long id, BytesRef dest) {
-        final long startOffset = startOffsets.get(id);
-        final int length = (int) (startOffsets.get(id + 1) - startOffset);
-        bytes.get(startOffset, length, dest);
-        return dest;
+    public BytesRefHash(final long initialCapacity, final float loadFactor, final Hasher hasher, final BigArrays bigArrays) {
+        assert initialCapacity > 0 : "initial capacity must be greater than 0";
+        assert loadFactor > 0 && loadFactor < 1 : "load factor must be between 0 and 1";
+
+        this.loadFactor = loadFactor;
+        this.hasher = hasher;
+        this.bigArrays = bigArrays;
+
+        capacity = Numbers.nextPowerOfTwo((long) (initialCapacity / loadFactor));
```

Comment:
```capacity``` becomes 64 when ```DEFAULT_INITIAL_CAPACITY``` is 32, is this intended?

Reply:
Yes, that's intended. Initial capacity of 32 is assumed with a load factor of 0.6, so we need at least 32/0.6 ~ 54 slots to begin with. The next power of two is 64. This also matches the existing implementation.